File: dart/lib/src/action/function_registry.dart
import '../ast/spec_ast.dart';
import '../trace/trace.dart';

final class UserFunctionRegistryException implements Exception {
  const UserFunctionRegistryException(this.message);

  final String message;

  @override
  String toString() => 'UserFunctionRegistryException: $message';
}

final class UserFunctionRegistry {
  const UserFunctionRegistry._({
    required this.entries,
    required Map<String, List<UserFunctionEntry>> byName,
  }) : _byName = byName;

  factory UserFunctionRegistry.fromSpec(
    SpecFile spec, {
    LinkedSpecTraceEmitter? trace,
  }) {
    return UserFunctionRegistry.fromFunctions(spec.functions, trace: trace);
  }

  factory UserFunctionRegistry.fromFunctions(
    Iterable<FunctionDefinition> functions, {
    LinkedSpecTraceEmitter? trace,
  }) {
    final functionList = functions.toList(growable: false);
    final traceScope = trace?.enterScope(
      'dart_compiler:function_registry',
      'function_count=${functionList.length}',
      LinkedSpecTraceLevel.high,
    );
    try {
      final entries = <UserFunctionEntry>[];
      final byName = <String, List<UserFunctionEntry>>{};

      var index = 0;
      for (final function in functionList) {
        final existing = byName[function.name];
        if (existing != null) {
          throw UserFunctionRegistryException(
            "duplicate user function '${function.name}'",
          );
        }
        final entry = UserFunctionEntry(index: index, definition: function);
        entries.add(entry);
        byName[function.name] = [entry];
        trace?.traceDecision(
          'dart_compiler:function_registry:definition',
          true,
          'index=$index name=${function.name} arity=${function.arity}',
          LinkedSpecTraceLevel.medium,
        );
        index += 1;
      }

      final registry = UserFunctionRegistry._(
        entries: List.unmodifiable(entries),
        byName: Map<String, List<UserFunctionEntry>>.unmodifiable({
          for (final item in byName.entries)
            item.key: List<UserFunctionEntry>.unmodifiable(item.value),
        }),
      );
      if (traceScope != null) {
        trace?.exitScope(traceScope, 'ok function_count=${entries.length}');
      }
      return registry;
    } on Object catch (error) {
      if (traceScope != null) {
        trace?.exitScope(traceScope, 'error=$error');
      }
      rethrow;
    }
  }

  static const empty = UserFunctionRegistry._(
    entries: [],
    byName: <String, List<UserFunctionEntry>>{},
  );

  final List<UserFunctionEntry> entries;
  final Map<String, List<UserFunctionEntry>> _byName;

  Iterable<String> get names => _byName.keys;

  List<StagedParseJob> get bodyParseJobs {
    return [
      for (final entry in entries)
        if (entry.bodyParseJob != null) entry.bodyParseJob!,
    ];
  }

  bool hasName(String name) => _byName.containsKey(name);

  List<int> expectedAritiesFor(String name) {
    final entries = _byName[name];
    if (entries == null) {
      return const [];
    }
    return List.unmodifiable([for (final entry in entries) entry.arity]);
  }

  UserFunctionEntry? lookup(String name) {
    final entries = _byName[name];
    if (entries == null || entries.isEmpty) {
      return null;
    }
    return entries.single;
  }

  UserFunctionEntry? resolveExact(String name, int arity) {
    final entries = _byName[name];
    if (entries == null) {
      return null;
    }
    for (final entry in entries) {
      if (entry.arity == arity) {
        return entry;
      }
      if (entry.signature != null && entry.acceptsArity(arity)) {
        return entry;
      }
    }
    return null;
  }

  UserFunctionCallResolution resolveCall(String name, int arity) {
    final entries = _byName[name];
    if (entries == null) {
      return UserFunctionCallResolution._(
        name: name,
        requestedArity: arity,
        expectedArities: const [],
        expectedArityDescriptions: const [],
      );
    }
    for (final entry in entries) {
      if (entry.acceptsArity(arity)) {
        return UserFunctionCallResolution._(
          name: name,
          requestedArity: arity,
          expectedArities: expectedAritiesFor(name),
          expectedArityDescriptions: [
            for (final candidate in entries) candidate.arityExpectation,
          ],
          entry: entry,
        );
      }
    }
    return UserFunctionCallResolution._(
      name: name,
      requestedArity: arity,
      expectedArities: expectedAritiesFor(name),
      expectedArityDescriptions: [
        for (final candidate in entries) candidate.arityExpectation,
      ],
    );
  }

  JsonObject toJson() {
    return {
      'functions': [for (final entry in entries) entry.toJson()],
      'body_parse_jobs': [for (final job in bodyParseJobs) job.toJson()],
    };
  }
}

final class UserFunctionEntry {
  const UserFunctionEntry({required this.index, required this.definition});

  final int index;
  final FunctionDefinition definition;

  String get name => definition.name;
  List<String> get params => definition.params;
  int get arity => definition.arity;
  Map<String, String> get parameterKinds => definition.parameterKinds;
  CallableSignature? get signature => definition.signature;
  String get bodySource => definition.bodySource;
  Object? get bodyPayload => definition.bodyPayload;
  StagedParseJob? get bodyParseJob => definition.bodyParseJob;
  Object? get bodyAst => definition.bodyAst;
  SourceSpan get sourceSpan => definition.sourceSpan;
  SourceSpan get bodySpan => definition.bodySpan;

  bool acceptsArity(int actual) {
    final callable = signature;
    if (callable == null) {
      return actual == arity;
    }
    return actual >= callable.minArity &&
        (callable.maxArity == null || actual <= callable.maxArity!);
  }

  String get arityExpectation {
    final callable = signature;
    if (callable == null) {
      return '$arity';
    }
    return callable.maxArity == null
        ? 'at least ${callable.minArity}'
        : 'exactly ${callable.minArity}';
  }

  JsonObject toJson() {
    return {'index': index, ...definition.toJson()};
  }

  JsonObject toDescriptorJson() {
    return {
      'index': index,
      'kind': 'user_function_definition',
      'version': parameterKinds.isNotEmpty ? 3 : (signature == null ? 1 : 2),
      'name': name,
      if (signature == null) 'params': params,
      if (signature == null) 'arity': arity,
      if (parameterKinds.isNotEmpty) 'parameter_kinds': parameterKinds,
      if (signature != null) 'signature': signature!.toJson(),
      'source_text': definition.source,
      'source_span': sourceSpan.toJson(),
      'body_span': bodySpan.toJson(),
      'body_source': bodySource,
      if (bodyPayload != null) 'body_payload': bodyPayload,
      if (bodyParseJob != null) 'body_parse_job': bodyParseJob!.toJson(),
      if (bodyAst != null) 'body_ast': bodyAst,
    };
  }
}

final class UserFunctionCallResolution {
  const UserFunctionCallResolution._({
    required this.name,
    required this.requestedArity,
    required this.expectedArities,
    required this.expectedArityDescriptions,
    this.entry,
  });

  final String name;
  final int requestedArity;
  final List<int> expectedArities;
  final List<String> expectedArityDescriptions;
  final UserFunctionEntry? entry;

  bool get nameKnown => expectedArities.isNotEmpty;
  bool get matched => entry != null;
  bool get arityMismatch => nameKnown && entry == null;
}

File: julia/src/action/FunctionRegistry.jl
struct UserFunctionRegistryException <: Exception
    message::String
end

Base.showerror(io::IO, error::UserFunctionRegistryException) = print(io, error.message)

struct UserFunctionEntry
    index::Int
    definition::FunctionDefinition
end

struct UserFunctionCallResolution
    name::String
    requested_arity::Int
    expected_arities::Vector{Int}
    entry::Union{Nothing,UserFunctionEntry}
    name_known::Bool
    matched::Bool
    arity_mismatch::Bool
end

function UserFunctionCallResolution(; name, requested_arity, expected_arities, entry = nothing)
    expected_vec = Int[expected_arities...]
    matched = entry !== nothing
    name_known = !isempty(expected_vec)
    return UserFunctionCallResolution(
        String(name),
        requested_arity,
        expected_vec,
        entry,
        name_known,
        matched,
        name_known && !matched,
    )
end

struct UserFunctionRegistry
    entries::Vector{UserFunctionEntry}
    by_name::Dict{String,Vector{UserFunctionEntry}}
end

function user_function_registry_from_spec(spec::SpecFile)
    return user_function_registry_from_functions(spec.functions)
end

function user_function_registry_from_functions(functions)
    entries = UserFunctionEntry[]
    by_name = Dict{String,Vector{UserFunctionEntry}}()

    for (index, definition) in enumerate(functions)
        name = definition.name
        if haskey(by_name, name)
            throw(UserFunctionRegistryException("duplicate user function '$name'"))
        end
        entry = UserFunctionEntry(index - 1, definition)
        push!(entries, entry)
        by_name[name] = UserFunctionEntry[entry]
    end

    return UserFunctionRegistry(entries, by_name)
end

empty_user_function_registry() = UserFunctionRegistry(UserFunctionEntry[], Dict{String,Vector{UserFunctionEntry}}())

function user_function_names(registry::UserFunctionRegistry)
    return [entry.definition.name for entry in registry.entries]
end

function body_parse_jobs(registry::UserFunctionRegistry)
    jobs = StagedParseJob[]
    for entry in registry.entries
        job = entry.definition.body_parse_job
        if job !== nothing
            push!(jobs, job)
        end
    end
    return jobs
end

has_user_function_name(registry::UserFunctionRegistry, name::AbstractString) = haskey(registry.by_name, String(name))

function expected_arities_for(registry::UserFunctionRegistry, name::AbstractString)
    entries = get(registry.by_name, String(name), nothing)
    if entries === nothing
        return Int[]
    end
    return [entry.definition.arity for entry in entries]
end

function lookup_user_function(registry::UserFunctionRegistry, name::AbstractString)
    entries = get(registry.by_name, String(name), nothing)
    if entries === nothing || isempty(entries)
        return nothing
    end
    return only(entries)
end

function resolve_exact_user_function(registry::UserFunctionRegistry, name::AbstractString, arity::Int)
    entries = get(registry.by_name, String(name), nothing)
    if entries === nothing
        return nothing
    end
    for entry in entries
        if user_function_accepts_arity(entry, arity)
            return entry
        end
    end
    return nothing
end

function user_function_accepts_arity(entry::UserFunctionEntry, arity::Int)
    signature = entry.definition.signature
    return signature === nothing ? arity == entry.definition.arity : arity >= signature.min_arity
end

function user_function_arity_expectation(entry::UserFunctionEntry)
    signature = entry.definition.signature
    return signature === nothing ? string(entry.definition.arity) : "at least $(signature.min_arity)"
end

function resolve_user_function_call(registry::UserFunctionRegistry, name::AbstractString, arity::Int)
    expected = expected_arities_for(registry, name)
    if isempty(expected)
        return UserFunctionCallResolution(name = name, requested_arity = arity, expected_arities = Int[])
    end
    entry = resolve_exact_user_function(registry, name, arity)
    return UserFunctionCallResolution(
        name = name,
        requested_arity = arity,
        expected_arities = expected,
        entry = entry,
    )
end

function function_definition_with_body_ast(definition::FunctionDefinition, body_ast)
    return FunctionDefinition(
        name = definition.name,
        params = definition.params,
        arity = definition.arity,
        signature = definition.signature,
        body_source = definition.body_source,
        body_payload = definition.body_payload,
        body_parse_job = definition.body_parse_job,
        body_ast = body_ast,
        source = definition.source,
        source_span = definition.source_span,
        body_span = definition.body_span,
    )
end

function stitch_function_body_ast(spec::SpecFile, job_id::AbstractString, body_ast)
    functions = FunctionDefinition[]
    found = false
    for definition in spec.functions
        job = definition.body_parse_job
        if job !== nothing && job.job_id == job_id
            if job.result_policy != "replace_field" || job.result_field != "body_ast"
                throw(UserFunctionRegistryException(
                    "function body parse job '$job_id' cannot stitch into body_ast",
                ))
            end
            push!(functions, function_definition_with_body_ast(definition, body_ast))
            found = true
        else
            push!(functions, definition)
        end
    end
    if !found
        throw(UserFunctionRegistryException("function body parse job '$job_id' not found"))
    end
    return SpecFile(functions = functions, rules = spec.rules)
end

function to_json(registry::UserFunctionRegistry)
    return Dict{String,Any}(
        "functions" => [to_json(entry) for entry in registry.entries],
        "body_parse_jobs" => [to_json(job) for job in body_parse_jobs(registry)],
    )
end

function to_json(entry::UserFunctionEntry)
    result = to_json(entry.definition)
    result["index"] = entry.index
    return result
end

function to_descriptor_json(entry::UserFunctionEntry)
    definition = entry.definition
    result = Dict{String,Any}(
        "index" => entry.index,
        "kind" => "user_function_definition",
        "version" => definition.signature === nothing ? 1 : 2,
        "name" => definition.name,
        "source_text" => definition.source,
        "source_span" => to_json(definition.source_span),
        "body_span" => to_json(definition.body_span),
        "body_source" => definition.body_source,
    )
    if definition.signature === nothing
        result["params"] = definition.params
        result["arity"] = definition.arity
    else
        result["signature"] = to_json(definition.signature)
    end
    _put_if_present!(result, "body_payload", definition.body_payload)
    if definition.body_parse_job !== nothing
        result["body_parse_job"] = to_json(definition.body_parse_job)
    end
    _put_if_present!(result, "body_ast", definition.body_ast)
    return result
end

function to_json(resolution::UserFunctionCallResolution)
    result = Dict{String,Any}(
        "name" => resolution.name,
        "requested_arity" => resolution.requested_arity,
        "expected_arities" => resolution.expected_arities,
        "name_known" => resolution.name_known,
        "matched" => resolution.matched,
        "arity_mismatch" => resolution.arity_mismatch,
    )
    if resolution.entry !== nothing
        result["entry"] = to_json(resolution.entry)
    end
    return result
end
